### ISSUE TITLE
How to get gradients using the C# API

### ISSUE THREAD

**Issue by mdabros** [Microsoft/CNTK#2661]
In the C++ and Python API there are forward and backward methods on the Function class, to handle forward and backward propagation. In the C# API there is only an Evaluate method, which I guess is closely related to the forward method. Is there any way of getting/calculating the gradients using the C# API?

There is a nice example of doing this in python:
```python
# compute the value and the derivative of the sigmoid at 0
v = C.input_variable(shape=(1,), needs_gradient=True)
f = C.sigmoid(v)
df, fv = f.forward({v:[[0]]}, [f.output], set([f.output]))
value = list(fv.values())[0]
grad = f.backward(df, {f.output: np.ones_like(value)}, set([v]))

value:
array([[ 0.5]], dtype=float32)
list(grad.values())[0]:
array([[ 0.25]], dtype=float32)`
```
Maybe someone could provide an example of achieving the same using C#?

**Comment by cesarsouza** [comment]
Indeed, is there any method analogous to cntk.ops.functions.Function.grad available from C#?

**Comment by floAr** [comment]
Bumping this up. Is there any plan to implement gradients in C# or is it not on the roadmap?

**Comment by YanYas** [comment]
Likewise, for work I'm trying to complete this feature is key. I find the lack of feedback on these subjects a little concerning. #3332 pertains to the same issue and it also has gone without explanation as to why this functionality hasn't been exposed in C#. Without an answer its hard to know if we are wasting our time and should we look elsewhere.

**Comment by miroslav22** [comment]
+1 This feature is vital and I can't use CNTK in a project because of this (and other missing API calls).  I haven't got time to try and add them in and compile it myself.

It's a shame that the C# API appears to be being neglected.  It surely can't be much work to interface the missing API entry points.
